src/Symfony/Bridge/Doctrine/Middleware/Debug/Connection.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bridge\Doctrine\Middleware\Debug;
+
+use Doctrine\DBAL\Driver\Connection as ConnectionInterface;
+use Doctrine\DBAL\Driver\Middleware\AbstractConnectionMiddleware;
+use Doctrine\DBAL\Driver\Result;
+use Doctrine\DBAL\Driver\Statement as DriverStatement;
+use Symfony\Component\Stopwatch\Stopwatch;
+
+/**
+ * @author Laurent VOULLEMIER <laurent.voullemier@gmail.com>
+ *
+ * @internal
+ */
+class Connection extends AbstractConnectionMiddleware
+{
+    private int $nestingLevel = 0;
+
+    public function __construct(
+        ConnectionInterface $connection,
+        private DebugDataHolder $debugDataHolder,
+        private ?Stopwatch $stopwatch,
+        private string $connectionName,
+    ) {
+        parent::__construct($connection);
+    }
+
+    public function prepare(string $sql): DriverStatement
+    {
+        return new Statement(
+            parent::prepare($sql),
+            $this->debugDataHolder,
+            $this->connectionName,
+            $sql,
+        );
+    }
+
+    public function query(string $sql): Result
+    {
+        $this->debugDataHolder->addQuery($this->connectionName, $query = new Query($sql));
+
+        $this->stopwatch?->start('doctrine', 'doctrine');
+        $query->start();
+        $result = parent::query($sql);
+        $query->stop();
+        $this->stopwatch?->stop('doctrine');
```

Comment:
```suggestion
        try {
            $result = parent::query($sql);
        } finally {
            $query->stop();
            $this->stopwatch?->stop('doctrine');
        }
```

Same for most of the other decorated functions. We need to be aware that any operation could potentially fail.

Reply:
Indeed, try/finally added, thank you :)